Parser/Token.cs

```diff
@@ -38,9 +39,37 @@ public enum E {
    // Utility function used to echo an error to the console
    public void PrintError () {
       if (Kind != ERROR) throw new Exception ("PrintError called on a non-error token");
-      Console.ForegroundColor = ConsoleColor.Yellow;
-      Console.WriteLine ($"At line {Line}, column {Column} of {Source.FileName}: {Text}");
-      Console.ResetColor ();
+      OutputEncoding = new UnicodeEncoding ();
+      var file = $"File: {Source.FileName}";
+      WriteLine (file);
+      WriteLine (new string ('\u2500', file.Length));
+
+      if (Line > 2) {
+         var (prevText, prevText1) = (Source.Lines[Line - 3], Source.Lines[Line - 2]);
+         WriteLine (string.Format ("{0,4}{1,0}{2,0}", Line - 2, "\u2502", prevText));
+         WriteLine (string.Format ("{0,4}{1,0}{2,0}", Line - 1, "\u2502", prevText1));
+      }
+      if (Line == 1)
+         WriteLine (string.Format ("{0,4}{1,0}{2,0}", Line - 1, "\u2502", Source.Lines[Line - 1]));
+
+      WriteLine (string.Format ("{0,4}{1,0}{2,0}", Line, "\u2502", Source.Lines[Line - 1]));

```

Comment:
There is a lot of repeated code here. Simplify and clean up. Assume that I change the spec to require 5 lines of context before and 5 lines of context after. Your code will look very clumsy. Try to write in the form of a loop so you can change the number of lines of context by just changing a parameter.